Add circuit for RETURNDATASIZE opcode

xiaodino opened the pull request.

xiaodino committed:
Add circuit for RETURNDATASIZE opcode

xiaodino committed:
Add circuit for RETURNDATASIZE opcode

CPerezz commented on bus-mapping/src/evm/opcodes/returndatasize.rs:
```diff
@@ -0,0 +1,112 @@
+    use pretty_assertions::assert_eq;
+
+    #[test]
+    fn returndatasize_opcode_impl() {
+        let code = bytecode! {
+            RETURNDATASIZE
+            STOP
+        };
```
I think this could be a bit more elaborated. 
We should create a sub-context with either: [CALL](https://www.evm.codes/#F1), [CALLCODE](https://www.evm.codes/#F2), [DELEGATECALL](https://www.evm.codes/#F4) or [STATICCALL](https://www.evm.codes/#FA) and then get a `RETURNDATASIZE` from there. I'd also try to play with really big values and values like 0 or similar. So that we test all the possible edge-cases.

xiaodino replied:
Thanks. I'm working on it.

xiaodino replied:
Updated the tests. Create a sub-context with ```CALL``` and then get a ```RETURNDATASIZE``` from there.

CPerezz commented on bus-mapping/src/evm/opcodes/returndatasize.rs:
```diff
@@ -0,0 +1,112 @@
+            None,
+            account_0_code_account_1_no_code(code),
+            tx_from_1_to_0,
+            |block, _tx| block.number(0xcafeu64),
```
There's no call data even in the tx here. 
So as said avobe this could be a bit more curated.

CPerezz commented on zkevm-circuits/src/evm_circuit/execution.rs:
```diff
@@ -448,7 +450,9 @@ impl<F: Field> ExecutionConfig<F> {
             sar_gadget: configure_gadget!(),
             extcodesize_gadget: configure_gadget!(),
             extcodecopy_gadget: configure_gadget!(),
+
             returndatasize_gadget: configure_gadget!(),
+
```
I'd remove these two empty lines added.

xiaodino replied:
Removed.

CPerezz commented on zkevm-circuits/src/evm_circuit/execution/returndatasize.rs:
```diff
@@ -0,0 +1,186 @@
+            Some(
+                return_data_size.to_le_bytes()[..N_BYTES_MEMORY_ADDRESS]
+                    .try_into()
+                    .unwrap(),
```
It's fine if we want to panic here. But I'd use `expect` and add a small description of the panic.

xiaodino replied:
Updated.

CPerezz commented on zkevm-circuits/src/evm_circuit/execution/returndatasize.rs:
```diff
@@ -0,0 +1,186 @@
+    use itertools::Itertools;
+    use mock::TestContext;
+
+    fn test_ok(return_data_size: usize, is_root: bool) {
```
As said before, I'd try to test for edge cases as really big returndata size numbers or 0 or things like this.
Otherwise, the testing is not as useful as it could be.

xiaodino replied:
Updated tests.

CPerezz reviewed (changes requested):
Great work! And thanks for submitting a PR for it. 

I have some things that should probably be addressed. Specifically with the testing cases applied here.

Aside from that the rest looks good! :smile:

xiaodino committed:
Update tests

xiaodino committed:
Update tests

smtmfft committed:
update last call info in return (#2)

Signed-off-by: smtmfft <smtm@taiko.xyz>

Signed-off-by: smtmfft <smtm@taiko.xyz>

smtmfft committed:
update last call info in return (#2)

Signed-off-by: smtmfft <smtm@taiko.xyz>

Signed-off-by: smtmfft <smtm@taiko.xyz>

xiaodino committed:
Update tests

xiaodino committed:
Update tests

xiaodino committed:
Merge branch 'xiaodino/returndatasize' of github.com:xiaodino/zkevm-circuits into xiaodino/returndatasize

xiaodino committed:
Update tests

xiaodino committed:
Update tests

xiaodino committed:
Update format in zkevm-circuits/src/evm_circuit/execution.rs

xiaodino committed:
Update format in zkevm-circuits/src/evm_circuit/execution.rs

xiaodino committed:
Change unwrap to expect in returndatasize.rs

xiaodino committed:
Change unwrap to expect in returndatasize.rs

xiaodino reviewed (commented):
@CPerezz  Thanks for the feedbacks. They have been addressed.

CPerezz reviewed (approved):
LGTM!! Thanks for your work on the comments @xiaodino 

Requesting one more review before merging.
Maybe @roynalnaruto can check this?

roynalnaruto commented:
I'll review this PR before end of today.

CPerezz commented:
@xiaodino please can you fix the issues with `rustfmt`?

smtmfft committed:
according to RETURN spec, callee info update happens only in case C, (#3)

i.e., only if it is not root call.

RETURN spec:

The `RETURN` opcode terminates the call successfully with return data for the
caller.

It behaves differently in different scenarios:

- `is_create` and `is_root`
    - A. Returns the specified memory chunk as deployment code.
    - B. End the execution
- `is_create` and `not is_root`
    - A. Returns the specified memory chunk as deployment code.
    - C. Restores caller's context and switch to it.
- `not is_create` and `is_root`
    - B. End the execution
- `not is_create` and `not is_root`
    - D. Returns the specified memory chunk to the caller.
    - C. Restores caller's context and switch to it.

Signed-off-by: smtmfft <smtmfft@taikocha.in>

Signed-off-by: smtmfft <smtmfft@taikocha.in>
Co-authored-by: smtmfft <smtmfft@taikocha.in>

xiaodino committed:
Merge branch 'xiaodino/returndatasize' of github.com:xiaodino/zkevm-circuits into xiaodino/returndatasize

xiaodino committed:
Fix the issues with rustfmt

xiaodino committed:
Fix merge conflicts

xiaodino commented:
> 

The issues with rustfmt have been fixed.

xiaodino committed:
Merge branch 'main' into xiaodino/returndatasize

xiaodino committed:
Update

xiaodino committed:
Merge branch 'xiaodino/returndatasize' of github.com:xiaodino/zkevm-circuits into xiaodino/returndatasize

xiaodino committed:
Add tests

roynalnaruto committed:
Merge branch 'main' into xiaodino/returndatasize

roynalnaruto committed:
Merge branch 'main' into xiaodino/returndatasize

CPerezz committed:
Merge branch 'main' into xiaodino/returndatasize

roynalnaruto committed:
Merge branch 'main' into xiaodino/returndatasize

roynalnaruto committed:
Merge branch 'main' into xiaodino/returndatasize

CPerezz committed:
Merge branch 'main' into xiaodino/returndatasize

xiaodino committed:
Resolve conflicts

xiaodino committed:
Merge branch 'xiaodino/returndatasize' of github.com:xiaodino/zkevm-circuits into xiaodino/returndatasize

xiaodino committed:
Resolve merge conflicts

ChihChengLiang commented:
@roynalnaruto is the review of this pr blocked?

roynalnaruto committed:
Merge branch 'main' into xiaodino/returndatasize

roynalnaruto commented:
> @roynalnaruto is the review of this pr blocked?

No, I was just busy working on the exponentiation feature and completely forgot about this. Will review by tomorrow end :)

roynalnaruto commented on zkevm-circuits/src/evm_circuit/execution/returndatasize.rs:
```diff
@@ -0,0 +1,186 @@
+#[derive(Clone, Debug)]
+pub(crate) struct ReturnDataSizeGadget<F> {
+    same_context: SameContextGadget<F>,
+    return_data_size: RandomLinearCombination<F, N_BYTES_MEMORY_ADDRESS>,
```
```suggestion
    return_data_size: RandomLinearCombination<F, N_BYTES_U64>,
```

Geth uses `uint64` for the length of return data:
https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L318

I think we should use `N_BYTES_U64`

xiaodino replied:
Make sense. Updated.

CPerezz replied:
I think that it should not be really necessary. As something bigger than U64 should imply that the memory expansion cost would be inassumible. 

And, so, we save some RLC saves for example.

roynalnaruto commented on zkevm-circuits/src/evm_circuit/execution/returndatasize.rs:
```diff
@@ -0,0 +1,186 @@
+            region,
+            offset,
+            Some(
+                return_data_size.to_le_bytes()[..N_BYTES_MEMORY_ADDRESS]
```
```suggestion
                return_data_size.to_le_bytes()[..N_BYTES_U64]
```

And the corresponding change here as well

xiaodino replied:
Updated.

CPerezz replied:
ditto

roynalnaruto commented on zkevm-circuits/src/evm_circuit/execution/returndatasize.rs:
```diff
@@ -0,0 +1,186 @@
+            RETURNDATASIZE
+            STOP
+        };
+        let ctx = TestContext::<2, 1>::new(
+            None,
+            |accs| {
+                accs[0].address(addr_b).code(code);
+                accs[1].address(addr_a).balance(Word::from(1u64 << 30));
+            },
+            |mut txs, accs| {
+                txs[0].to(accs[0].address).from(accs[1].address);
+            },
+            |block, _tx| block,
+        )
+        .unwrap();
+
+        assert_eq!(run_test_circuits(ctx, None), Ok(()));
+    }
```
Not a biggie, but I think you can simplify this right? Or am I missing something?

```suggestion
    assert_eq!(
        run_test_circuits(
            TestContext::<2, 1>::simple_ctx_with_bytecode(code).unwrap(),
            None              
        ),                                                                                                                 
        Ok(())
    );
```

xiaodino replied:
Updated. Thanks.

roynalnaruto reviewed (commented):
Overall looks good! A few comments, a few suggestions (especially the change from `N_BYTES_MEMORY_ADDRESS` to `N_BYTES_U64`) and a few questions.

Good work! And thanks for contributing :)

xiaodino committed:
Merge branch 'main' into xiaodino/returndatasize

xiaodino committed:
Update bus-mapping return.rs, use N_BYTES_U64 for return_data_size and update tests

xiaodino committed:
Merge branch 'xiaodino/returndatasize' of github.com:xiaodino/zkevm-circuits into xiaodino/returndatasize

xiaodino commented on bus-mapping/src/evm/opcodes/return.rs:
```diff
@@ -42,6 +43,27 @@ impl Opcode for Return {
+
+            state.call_context_write(
+                &mut exec_step,
+                state.call()?.call_id,
```
Good catch. Updated.

Thanks.

CPerezz commented:
@roynalnaruto I actually don't really agree with the U256 instead of U64 used for the value RLC.

See: https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/703#discussion_r968742152

roynalnaruto commented:
> @roynalnaruto I actually don't really agree with the U256 instead of U64 used for the value RLC.
> 
> See: [#703 (comment)](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/703#discussion_r968742152)

I mentioned that since geth uses `uint64`, so I thought we should stick to 8 bytes (`N_BYTES_U64`). `N_BYTES_MEMORY_ADDRESS` was 5 bytes.

Brechtpd commented:
I think in most opcodes we do use `N_BYTES_MEMORY_WORD_SIZE` for storing a similar value (memory length/memory offsets in `MemoryAddress` types). Though a small difference here is that this opcode on its own does not not need any memory expansion (which is where the gas cost would overflow a 64bit value for these 5-byte sizes/offsets already) because it's a simple getter. So in that aspect `N_BYTES_U64` could make sense instead, but of course the `RETURN` will already have done the memory expansion so it is as impossible to have bigger values than `N_BYTES_MEMORY_WORD_SIZE` as in all other uses cases where that is used I believe.

So good points to make for both I feel, but practically `N_BYTES_MEMORY_WORD_SIZE` should be as okay as its use in other opcodes. On the other hand, using `N_BYTES_U64` won't make that much difference.

z2trillion committed:
Merge branch 'main' into xiaodino/returndatasize

roynalnaruto committed:
Merge branch 'main' into xiaodino/returndatasize

roynalnaruto committed:
Merge branch 'main' into xiaodino/returndatasize

han0110 commented on bus-mapping/src/evm/opcodes/return.rs:
```diff
@@ -42,6 +43,27 @@ impl Opcode for Return {
                 let code = memory.0[offset..offset + length].to_vec();
                 state.code_db.insert(code);
             }
+
+            state.call_context_write(
+                &mut exec_step,
+                state.caller()?.call_id,
+                CallContextField::LastCalleeId,
+                state.call()?.call_id.into(),
+            );
+
+            state.call_context_write(
+                &mut exec_step,
+                state.caller()?.call_id,
+                CallContextField::LastCalleeReturnDataOffset,
+                offset.into(),
+            );
+
+            state.call_context_write(
+                &mut exec_step,
+                state.caller()?.call_id,
+                CallContextField::LastCalleeReturnDataLength,
+                length.into(),
+            );
```
After rebase on `main`, these should be handled in `handle_restore_context` and could be removed.

xiaodino replied:
Is this PR: https://github.com/xiaodino/zkevm-circuits/commit/809895b75630c6ee42eadfba85e8274389a20c5b the one that handle `state.call_context_write`?

xiaodino replied:
Updated. Please take a look

han0110 reviewed (commented):
LGTM! After rebase we could merge this.

xiaodino committed:
Revert changes in bus-mapping return.rs

xiaodino committed:
Merge branch 'xiaodino/returndatasize' of github.com:xiaodino/zkevm-circuits into xiaodino/returndatasize

xiaodino committed:
Revert change in bus-mapping return

han0110 reviewed (approved):
LGTM! Once the CI passes we can merge this.

adria0 committed:
Merge branch 'main' into xiaodino/returndatasize

xiaodino committed:
Merge branch 'main' into xiaodino/returndatasize

xiaodino committed:
Update bus-mapping returndatasize_tests

xiaodino committed:
Merge branch 'xiaodino/returndatasize' of github.com:xiaodino/zkevm-circuits into xiaodino/returndatasize

xiaodino commented:
> LGTM! Once the CI passes we can merge this.

CI passed!
